src/main/java/org/apache/mesos/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,311 @@
+                    acceptedOffers = planScheduler.resourceOffers(driver, offers, block.get());
+                    logger.info(String.format("Accepted %d of %d offers: %s",
+                            acceptedOffers.size(), offers.size(), acceptedOffers));
+                }
+                List<Protos.Offer> unacceptedOffers = filterAcceptedOffers(offers, acceptedOffers);
+
+                // Recovery
+                try {
+                    acceptedOffers.addAll(recoveryScheduler.resourceOffers(driver, unacceptedOffers, block));
+                } catch (Exception e) {
+                    logger.error("Error repairing block: " + block + " Reason: " + e);
+                }
+
+                // Leaked resource recovery
+                ResourceCleanerScheduler cleanerScheduler = getCleanerScheduler();
+                if (cleanerScheduler != null) {
+                    acceptedOffers.addAll(getCleanerScheduler().resourceOffers(driver, offers));
+                }
+
+                declineOffers(driver, acceptedOffers, offers);
+
+                // Kill tasks (for configuration update, or user requested Task restart or replace)
+                taskKiller.process(driver);
+            }
+        });
+    }
+
+    @Override
+    public void offerRescinded(SchedulerDriver driver, Protos.OfferID offerId) {
+        logger.error("Rescinding offers is not supported.");
+        hardExit(SchedulerErrorCode.OFFER_RESCINDED);
+    }
+
+    @Override
+    public void statusUpdate(SchedulerDriver driver, Protos.TaskStatus status) {
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                logger.info(String.format(
+                        "Received status update for taskId=%s state=%s message='%s'",
+                        status.getTaskId().getValue(),
+                        status.getState().toString(),
+                        status.getMessage()));
+
+                // Store status, then pass status to PlanManager => Plan => Blocks
+                try {
+                    stateStore.storeStatus(status);
+                    planManager.update(status);
+                } catch (Exception e) {
+                    logger.warn("Failed to update TaskStatus received from Mesos. "
+                            + "This may be expected if Mesos sent stale status information: " + status, e);
+                }
+            }
+        });
+    }
+
+    @Override
+    public void frameworkMessage(
+            SchedulerDriver driver,
+            Protos.ExecutorID executorId,
+            Protos.SlaveID slaveId,
+            byte[] data) {
+        logger.error("Received a Framework Message, but don't know how to process it");
+        hardExit(SchedulerErrorCode.FRAMEWORK_MESSAGE);
+    }
+
+    @Override
+    public void disconnected(SchedulerDriver driver) {
+        logger.error("Disconnected from Master.");
+        hardExit(SchedulerErrorCode.DISCONNECTED);
```

Comment:
To avoid a scenario where it's possible that more than 1 Scheduler is doing anything.  Because the DefaultScheduler cannot do anything until it is registered and only one Scheduler may be registered at a time, we guarantee that no work is accidentally done by more than 1 Scheduler simultaneously by crashing on disconnect.